File: v3/household.py
import discord

import typing
import datetime


class Household:

    def __init__(self, name: str, channel: discord.TextChannel):
        self.name = name
        self.channel = channel
        self._chores = []
        self._users = set()
        self._offset = 0
        self._chore_list = []

    def add_user(self, user: discord.User) -> None:
        self._users.add(user)

    def remove_user(self, user: discord.User) -> None:
        self._users.remove(user)

    def get_users(self) -> set:
        return self._users

    def set_chores(self, chores: list):
        self._chores = chores

    def get_chores(self) -> list:
        return self._chores

    def get_active_chore(self) -> typing.Optional[dict]:
        if len(self._chore_list) == 0:
            return None
        chore_set = self._chore_list[-1]
        if datetime.datetime.now().date() < chore_set['end_date']:
            return None
        else:
            return chore_set

    def new_chore_set(self, end_date: datetime.datetime = None) -> dict:
        assignments = []
        for index, user in enumerate(self.get_users()):
            assignments.append({
                'user': user,
                'chore': self.get_chores()[(self.get_offset() + index) % len(self.get_chores())],
                'date-complete': None
            })
        self.advance_offset()
        chore_set = {
            'chore_message_id': None,
            'unfinished': assignments,
            'finished': [],
            'start_date': datetime.datetime.now().date(),
            'end_date': end_date
        }
        self._chore_list.append(chore_set)
        return chore_set

    def advance_offset(self, n: int = 1) -> None:
        self._offset = (self.get_offset() + n) % len(self.get_chores())

    def get_offset(self) -> int:
        return self._offset

    def complete(self, user: discord.User, date) -> bool:
        chore_set = self.get_active_chore()
        if not chore_set:
            return False

File: src/household.js
import Assignment from './assignment';
import ChoreBoard from './chore-board';

class Household {
  constructor(name, channel) {
    this.name = name;
    this.channel = channel;
    this.chores = [];
    this.users = [];
    this.offset = 0;
    this.choreBoardList = [];
    this.autoRenew = null;
  }

  addUser(user) {
    if (!this.users.includes(user)) {
      this.users.push(user);
    }
  }

  getUserById(userId) {
    return this.users.find(u => u.id === userId);
  }

  removeUser(user) {
    this.users = this.users.filter(item => item !== user);
  }

  setChores(chores) {
    this.chores = chores;
  }

  getActiveChoreBoard() {
    if (this.choreBoardList.length === 0) { // make sure there is at least one
      return null;
    }
    const choreBoard = this.choreBoardList[this.choreBoardList.length - 1]; // get most recent board
    if (new Date() > choreBoard.endDate) { // make sure it's still active
      return null;
    }
    return choreBoard;
  }

  newChoreBoard(endDate) {
    const assignments = [];
    this.users.forEach((user, index) => {
      assignments.push(new Assignment(
        user,
        this.chores[(this.offset + index) % this.chores.length]
      ));
    });
    this.advanceOffset(); // move offset for next assignment
    const choreBoard = new ChoreBoard(assignments, endDate); // create and return board
    this.choreBoardList.push(choreBoard);
    return choreBoard;
  }

  advanceOffset(n = 1) {
    this.offset = (this.offset + n) % this.chores.length;
  }

  complete(user, date) {
    const choreBoard = this.getActiveChoreBoard(); // check for active board
    if (!choreBoard) {
      return null;
    }
    for (let assignment of choreBoard.unfinishedAssignments) { // check for user in unfinished assignments
      if (assignment.user === user) {
        assignment.dateComplete = date; // mark and sort assignment as complete
        choreBoard.unfinishedAssignments = choreBoard.unfinishedAssignments.filter(a => a !== assignment);
        choreBoard.finishedAssignments.push(assignment);
        return choreBoard;
      }
    }
    return null;
  }
}

export default Household;
